### PR TITLE
Add datefilter directive and associated components

### PR DESCRIPTION
**Pull request by moradology**
This PR includes styling that should bring us much closer to the filterbar prototype

### CODE REVIEW

File: web/test/spec/filterbar/date-range-directive.spec.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+        $rootScope.$apply();
+    }));
+
+    it('should handle restoring filter selection', function () {
+        // should have no maximum set yet
+        expect(Element.find("input[type='text'][name='maximum']").val()).toEqual('');
+
+        // mine goes to eleven
+        $rootScope.$broadcast('driver.filterbar:restore', {'__dateRange': {min: 0, max: 11}});
```
Comment (flibbertigibbet) [inline web/test/spec/filterbar/date-range-directive.spec.js]: Maybe test this with an actual date object instead of a number?

File: web/app/scripts/state/filters-service.js [hunk 2]
```diff
@@ -16,29 +23,45 @@
...
+
 
         /**
          * Broadcast event to trigger setting previously stored filters back on the filter bar.
          */
-        svc.restoreFilters = function() {
+        function restoreFilters() {
             var filterString = localStorageService.get(storageName);
-            var filterObj = !!filterString ? JSON.parse(filterString) : {};
+            var filterObj = !!filterString ? filterString : {};
```
Comment (flibbertigibbet) [inline web/app/scripts/state/filters-service.js]: Why remove the `parse` here? Storage contains stringified objects.

File: web/app/scripts/state/filters-service.js [hunk 3]
```diff
@@ -16,29 +23,45 @@
          *
          * @param {Object} filters FilterBar's filters object to restore on load.
          */
-        svc.saveFilters = function(filters) {
-            localStorageService.set(storageName, JSON.stringify(filters));
-        };
+        function saveFilters(filters) {
+            localStorageService.set(storageName, filters);
```
Comment (flibbertigibbet) [inline web/app/scripts/state/filters-service.js]: Objects must be stringified to go into storage.

File: web/app/scripts/filterbar/numeric-range-directive.js [hunk 4]
```diff
@@ -15,17 +15,18 @@
...
                     scope.error = {};
-
-                    // restore previously set filter selection on page reload
-                    scope.$on('driver.filterbar:restored', function(event, filter) {
-                        if (filter.label === scope.label) {
-                            scope.filter = filter.value;
-                            scope.isMinMaxValid();
-                        }
-                    });
+                    scope.filter = FilterState.get(scope.label);
```
Comment (flibbertigibbet) [inline web/app/scripts/filterbar/numeric-range-directive.js]: I don't see the reason for the changes to this file. `FilterState` should not be referenced in the specific filter directives; that's managed at a higher level. The numeric range filters were working fine.

File: web/app/scripts/state/filters-service.js (hunk 2 above)
Comment (moradology) [inline web/app/scripts/state/filters-service.js]: The localStorageService already parses in its get method, so I think this was noop at best

File: web/app/scripts/filterbar/date-range-directive.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            templateUrl: 'scripts/filterbar/date-range.html',
+            controller: 'dateRangeController',
+            link: function(scope, elem, attrs, ctlArray) {
+                var filterLabel = '__dateRange';
+                init();
+                scope.calendarOptions = {'format': ctlArray[1].dateTimeFormat};
+
+                // restore previously set filter selection on page reload
+                scope.$on('driver.filterbar:restored', function(event, filter) {
+                    console.log(filter);
```
Comment (flibbertigibbet) [inline web/app/scripts/filterbar/date-range-directive.js]: Remove console log here, or use `$log`.

File: web/app/scripts/filterbar/date-range-directive.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    function DateRangeField(FilterState, $timeout) {
+        var module = {
+            restrict: 'A',
+            require: ['^driver-filterbar', 'date-range-field'],
+            templateUrl: 'scripts/filterbar/date-range.html',
+            controller: 'dateRangeController',
+            link: function(scope, elem, attrs, ctlArray) {
+                var filterLabel = '__dateRange';
+                init();
+                scope.calendarOptions = {'format': ctlArray[1].dateTimeFormat};
```
Comment (flibbertigibbet) [inline web/app/scripts/filterbar/date-range-directive.js]: It would be nice to locally define controller vars instead of referencing them by array offset, for clarity.

File: web/app/scripts/app.js [hunk 7]
```diff
@@ -62,5 +62,6 @@
     .config(LogConfig)
     .config(LeafletDefaultsConfig)
     .config(LocalStorageConfig)
-    .config(HttpConfig);
+    .config(HttpConfig)
+    .run(function(FilterState) { FilterState.restoreFilters(); });
```
Comment (flibbertigibbet) [inline web/app/scripts/app.js]: Why is this here?

File: web/app/scripts/state/filters-service.js (hunk 3 above)
Comment (moradology) [inline web/app/scripts/state/filters-service.js]: I think it automatically serializes as well

Comment (flibbertigibbet) [inline web/app/scripts/state/filters-service.js]: Ok, neat

File: web/app/scripts/app.js (hunk 7 above)
Comment (moradology) [inline web/app/scripts/app.js]: An oversight left from something I was testing out - removing

File: web/app/scripts/filterbar/date-range-directive.js (hunk 6 above)
Comment (moradology) [inline web/app/scripts/filterbar/date-range-directive.js]: That's a good idea

File: web/app/scripts/state/filters-service.js [hunk 8]
```diff
@@ -16,29 +23,45 @@
...
 
             $log.debug('Restoring filters:');
             $log.debug(filterObj);
 
             // tell the filterbar to set the filters back
             $rootScope.$broadcast('driver.filterbar:restore', filterObj);
-        };
+        }
 
+        svc.restoreFilters();
```
Comment (flibbertigibbet) [inline web/app/scripts/state/filters-service.js]: Why is this here?

File: web/bower.json [hunk 9]
```diff
@@ -24,7 +24,8 @@
     "Leaflet.utfgrid": "danzel/Leaflet.utfgrid",
     "d3": "3.5.5",
     "d3-tip": "^0.6.7",
-    "moment": "^2.10.6"
+    "moment": "^2.10.6",
+    "jquery": "^2.1.4"
```
Comment (flibbertigibbet) [inline web/bower.json]: What requires proper jQuery? This won't conflict with the Angular jQuery lite?

Comment (moradology) [inline web/bower.json]: I think angular can use jQuery rather than jqLite, but you're right that it is not necessary - angular-ui-bootstrap's implementation doesn't require jQuery like the datepicker otherwise would

**Comment by flibbertigibbet** [comment]
+1
